main/src/main/java/cgeo/geocaching/connector/gc/ImportBookmarkLinks.java

```diff
@@ -9,13 +9,90 @@
 import android.net.Uri;
 import android.os.Bundle;
 
+import androidx.annotation.NonNull;
 import androidx.annotation.Nullable;
 import androidx.appcompat.app.AppCompatActivity;
 
 import java.util.regex.Pattern;
 
+import javax.annotation.RegEx;
+
 public class ImportBookmarkLinks extends AppCompatActivity {
 
+    private final UrlToIdParser intentUrlParser = defaultBookmarkListUrlToIdParser();
+
+    public static final String BOOKMARK_LIST_API_PREFIX =
+            "https://www.geocaching.com/plan/api/gpx/list/";
+
+    interface UrlToIdParser {
+        @Nullable String tryExtractFromIntentUrl(@Nullable String intentUrl);
+    }
+
+    static class SingleUrlToIdParser implements UrlToIdParser {
+        private final Pattern matcherPattern;
+        // Android API 21 apparently does not support named capturing groups in Pattern
+        // which has introduced in java 1.7...
+        private final int groupToExtract;
+
+        SingleUrlToIdParser(
+                @RegEx @NonNull final String regex,
+                final int groupToExtract
+        ) {
+            this.matcherPattern = Pattern.compile(regex, Pattern.CASE_INSENSITIVE);
+            this.groupToExtract = groupToExtract;
+        }
+
+        @Override
+        public String tryExtractFromIntentUrl(final String intentUrl) {
+            if (intentUrl == null) {
+                return null;
+            }
+            final MatcherWrapper matcher = new MatcherWrapper(this.matcherPattern, intentUrl);
+            if (matcher.find()) {
+                final String maybeId = matcher.group(groupToExtract);
+                if (maybeId != null) {
+                    return maybeId.toUpperCase();
+                }
+            }
+            return null;
+        }
+    }
+
+    static class AggregatedUrlToIdParser implements UrlToIdParser {
+        private final UrlToIdParser[] parsers;
+
+        AggregatedUrlToIdParser(final UrlToIdParser... parsers) {
+            this.parsers = parsers;
+        }
+
+        @Override
+        public String tryExtractFromIntentUrl(final String intentUrl) {
+            if (intentUrl == null) {
+                return null;
+            }
+            for (final UrlToIdParser parser : parsers) {
+                final String id = parser.tryExtractFromIntentUrl(intentUrl);
+                if (id != null) {
+                    return id;
+                }
+            }
+            return null;
+        }
+    }
+
+    static UrlToIdParser defaultBookmarkListUrlToIdParser() {
+        return new AggregatedUrlToIdParser(
+                new SingleUrlToIdParser(
+                        "^https?://(?:www\\.)?geocaching\\.com/plan/lists/([A-Z0-9]+)",
```

Comment:
As stated above: it might make sense to use naming groups in the regex anyway and comment which group index corresponds to the naming group

Reply:
Done